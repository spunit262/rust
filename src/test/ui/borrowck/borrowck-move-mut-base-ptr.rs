// Test that attempt to move `&mut` pointer while pointee is borrowed
// yields an error.
//
// Example from src/librustc_borrowck/borrowck/README.md



fn foo(t0: &mut isize) {
    let p: &isize = &*t0; // Freezes `*t0`
    let t1 = t0; //~ ERROR cannot borrow `*t0` as mutable because it is also borrowed as immutable
    *t1 = 22;
    p.use_ref();
}

fn main() {
}

trait Fake { fn use_mut(&mut self) { } fn use_ref(&self) { }  }
impl<T> Fake for T { }
