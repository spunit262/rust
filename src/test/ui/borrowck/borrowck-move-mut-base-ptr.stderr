error[E0502]: cannot borrow `*t0` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-move-mut-base-ptr.rs:10:14
   |
LL |     let p: &isize = &*t0; // Freezes `*t0`
   |                     ---- immutable borrow occurs here
LL |     let t1 = t0;
   |              ^^ mutable borrow occurs here
LL |     *t1 = 22;
LL |     p.use_ref();
   |     - immutable borrow later used here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0502`.
